src/oauth/middleware.ts

```diff
@@ -19,29 +23,155 @@ export type Variables = {
+    if (credentials?.username && credentials.password) {
+      clientCredentials.push({
+        authentication: "client_secret_basic",
+        client_id: credentials.username,
+        client_secret: credentials.password,
+      });
+    }
+  }
+
+  // client authentication: client_secret_post
+  if (c.req.method === "POST") {
+    const body = await c.req.parseBody();
+    const result = await clientCredentialsSchema.safeParse(body);
+
+    if (result.success) {
+      clientCredentials.push({
+        authentication: "client_secret_post",
+        client_id: result.data.client_id,
+        client_secret: result.data.client_secret,
+      });
+    }
+  }
 
-  if (accessToken === undefined) {
-    return c.json({ error: "invalid_token" }, 401);
+  // client authentication: none
+  const client_id_param = c.req.query("client_id");
+  const client_secret_param = c.req.query("client_secret");
+  if (client_id_param && client_secret_param === undefined) {
+    clientCredentials.push({
+      authentication: "none",
+      client_id: client_id_param,
+      client_secret: undefined,
+    });
   }
 
-  c.set("token", accessToken);
+  if (clientCredentials.length > 1) {
+    return c.json(
+      {
+        error: "invalid_request",
+        error_description:
+          "The request includes includes multiple credentials or utilizes more than one mechanism for authenticating the client",
+      },
+      400,
+    );
+  }
+
+  if (clientCredentials.length === 0) {
+    return c.json(
+      {
+        error: "invalid_client",
+        error_description:
+          "Client authentication failed due to no client authentication included, or unsupported authentication method",
+      },
+      401,
+    );
+  }
+
+  let client: Application | undefined;
+  if (
+    clientCredentials[0].authentication === "client_secret_basic" ||
+    clientCredentials[0].authentication === "client_secret_post"
+  ) {
+    client = await db.query.applications.findFirst({
+      where: and(
+        eq(applications.clientId, clientCredentials[0].client_id),
+        eq(applications.clientSecret, clientCredentials[0].client_secret),
+        eq(applications.confidential, true),
+      ),
+    });
+  } else {
+    // client authentication method is none, which only works for non-confidential clients:
+    client = await db.query.applications.findFirst({
+      where: and(
+        eq(applications.clientId, clientCredentials[0].client_id),
+        eq(applications.confidential, false),
+      ),
+    });
+  }
+
+  if (!client) {
+    return c.json(
+      {
+        error: "invalid_client",
+        error_description:
+          "Client authentication failed due to unknown client, " +
+          "no client authentication included, or unsupported authentication " +
+          "method.",
+      },
+      401,
+    );
+  }
+
+  c.set("client", client);
   await next();
 });
 
+export const tokenRequired = createMiddleware<{ Variables: Variables }>(
```

Comment:
The only change I made here was to the `createMiddleware<{ Variables: Variables }>(` signature, other than that it just moved down in the file and the git diff obscures this fact.